packages/shell-api/src/streams.ts

```diff
@@ -0,0 +1,172 @@
+import type { Document } from 'bson';
+import { CommonErrors, MongoshInvalidInputError } from '@mongosh/errors';
+
+import { MongoshInterruptedError } from './interruptor';
+import {
+  ShellApiWithMongoClass,
+  returnsPromise,
+  shellApiClassDefault,
+} from './decorators';
+import type ShellInstanceState from './shell-instance-state';
+import StreamProcessor from './stream-processor';
+import { ADMIN_DB, asPrintable } from './enums';
+import type Mongo from './mongo';
+
+@shellApiClassDefault
+export class Streams extends ShellApiWithMongoClass {
+  public static newInstance(instanceState: ShellInstanceState) {
+    return new Proxy(new Streams(instanceState), {
+      get(target, prop) {
+        const v = (target as any)[prop];
+        if (v !== undefined) {
+          return v;
+        }
+        if (typeof prop === 'string') {
+          return target.getProcessor(prop);
+        }
+      },
+    });
+  }
+
+  constructor(private instanceState: ShellInstanceState) {
+    super();
+  }
+
+  get _mongo(): Mongo {
+    return this.instanceState.currentDb._mongo;
+  }
+
+  [asPrintable](): string {
+    return 'Atlas Stream Processing';
+  }
+
+  getProcessor(name: string) {
+    return new StreamProcessor(this, name);
+  }
+
+  @returnsPromise
+  async process(pipeline: Document[], options: Document = {}) {
+    if (!Array.isArray(pipeline) || !pipeline.length) {
+      throw new MongoshInvalidInputError(
+        'Invalid pipeline',
+        CommonErrors.InvalidArgument,
+        pipeline
+      );
+    }
+    const result = await this._runCommand({
+      processStreamProcessor: 1,
+      pipeline,
+      options,
+    });
+
+    if (result.ok !== 1) {
+      return result;
+    }
+
+    const { name, cursorId } = result as {
+      name: string;
+      limit: number;
+      cursorId: number;
+    };
+    const sp = this.getProcessor(name);
+
+    const stopAndDrop = () => {
+      sp.stop()
+        .then(() => sp.drop())
+        .catch(() => {
+          /* ignore */
+        });
+    };
+
+    try {
+      await sp._sampleFrom(cursorId);
+    } catch (err) {
+      // try to stop and drop the temp processor when interrupted
+      if (err instanceof MongoshInterruptedError) {
+        this._instanceState.messageBus.once(
+          'mongosh:interrupt-complete',
+          stopAndDrop
+        );
+      }
```

Comment:
As the documentation for `interrupt-complete` notes, it’s a testing-only event. Why do we need to wait for it here? Why do we not call `stopAndDrop()` for other errors? Should this just be a 

```js
try { 
  await sp._sampleFrom(cursorId);
} finally { 
 try { 
    await sp.stop();
    await sp.drop();
  } catch { /* ignore */ }
}
```

?

Reply:
For interrupted error specific, from memory, I think the call will always fail as the underlying connection is closed. So we need to wait until the interruption is handled. 
For other errors, I wasn't sure what other errors could occur. Only error I can think of is network issue. In which case calling the method will also fail. So, I was just leaving that for server to handle. But having said that, we can always call the method on error as you suggested.

Reply:
Thank you, that makes sense! Then one thing that should probably change here is that we’d want to run any code that should wait the connection to be re-established from `onResumeExecution` rather than in response to the `interrupt-complete` event (and in particular keep the control flow local to the shell-api package, instead of relying on the user of the package to always emit this event).

I’ll also try to think whether there is a better way to handle this situation in general – it might be nice to actually allow creating sections of code where the behavior on Ctrl+C is fully overriden, so that there wouldn’t be a connection reset happening in the first place.